hamt/partial-enum.md

```diff
@@ -0,0 +1,29 @@
+# Partial enumeration
+
+This a proposal to implement on top of https://github.com/ipfs/go-unixfs/pull/94 to allow the performance optimization of not having to enumerate *all* links in the HAMT directory when evaluating a possible downgrade/switch to the basic single-node directory.
+
+Note on terminology: "partial" may not be the most apt term here. It is just used as a contrast with the *full* enumeration of `ForEachLink` and `EnumLinksAsync`. The exact nature of the enumeration proposed needs to be further elucidated in the proposal here.
+
+*tl;dr*: keep fetched nodes from the enumeration in the `Shard` structure (as if they had been fetched as a consequence of an `AddChild`/`RemoveChild` call, which are retained throughout the lifetime of the structure/directory).
+
+## Trade memory for bandwidth
+
+The key insight of this proposal is the trade-off between memory (MEM) consumed by the `HAMTDirectory`/`Shard` and the bandwidth (BW) used to fetch the different nodes comprising the full `HAMTDirectory` DAG during enumeration.
```

Comment:
What memory are we trading off and why?

Aren't we just trying to minimize bandwidth by only fetching what we need and then we have the option of trading off BW vs latency by concurrently fetching more of the shards and perhaps overestimating the number we need to download?

Reply:
The memory of holding the `Shard` structure in the `childer.children` instead of the `links`:

https://github.com/ipfs/go-unixfs/blob/12b9a8676a9d0bbda1ea61cdc9c196c2132b8ff4/hamt/hamt.go#L540-L541